### PR TITLE
Issue #12625: validate wildcards on target classes

### LINKED ISSUE #12625: pitest: validate wildcards on target classes
Identified at https://github.com/checkstyle/checkstyle/pull/12622 ,

https://github.com/checkstyle/checkstyle/blob/811983e0f89edf2419b9165ca60bb04a7c3c1ea6/pom.xml#L3355

We need to create a new validation, that ensures all `targetClasses` provided to the pitest profiles have a `*` on the end of them. The `*` is needed to include all sub-classes of the target class. Without it, pitest will miss mutations as we saw in the mentioned PR.

### PR DESCRIPTION
**Pull request by nrmancuso**
Closes #12625

Link to example of failed run: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/16973/workflows/1abde372-b58a-4607-9bb5-a51560e8edf2/jobs/231413/parallel-runs/0/steps/0-102 

```
Append asterisks to the following pitest target classes in pom.xml:
com.puppycrawl.tools.checkstyle.AuditEventFormatter
com.puppycrawl.tools.checkstyle.XdocsPropertyType
com.puppycrawl.tools.checkstyle.PropertyType
com.puppycrawl.tools.checkstyle.FileStatefulCheck
com.puppycrawl.tools.checkstyle.ModuleFactory
com.puppycrawl.tools.checkstyle.PropertyResolver
com.puppycrawl.tools.checkstyle.StatelessCheck
com.puppycrawl.tools.checkstyle.TreeWalkerFilter
com.puppycrawl.tools.checkstyle.GlobalStatefulCheck
com.puppycrawl.tools.checkstyle.grammar.CommentListener


Exited with code exit status 10
```

### CODE REVIEW

**Review by romani — APPROVED** [review]
ok to merge

**Comment by nrmancuso** [comment]
I am dealing with strange apt issue, needs some work here still.

**Comment by nrmancuso** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
GitHub, generate report

**Comment by nrmancuso** [comment]
GitHub, generate report

**Comment by nrmancuso** [comment]
I wonder if failure is related to https://github.com/checkstyle/contribution/commit/232c89063dc398a932534d1296f170dc25f413c5

**Comment by nrmancuso** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
Diff report generation has nothing to do with this PR, I only used it for testing to not waste CI run in test PR. 

ATTENTION: watch for old [doctype](https://github.com/checkstyle/contribution/commit/6a8a86ca685e1a45359ccaeb890465058026992d) in config XML, diff report generation will fail without it. 

Failure looks like:
```
Starting diff report generation ...
Running command: java -jar /home/runner/work/checkstyle/checkstyle/contribution/patch-diff-report-tool/target/patch-diff-report-tool-0.1-SNAPSHOT-jar-with-dependencies.jar --patchReport reports/origin/issue-12625/pmd/checkstyle-result.xml
                        --output reports/diff/pmd --patchConfig patch_config.xml --baseReport reports/upstream/master/pmd/checkstyle-result.xml --baseConfig diff_config.xml
patch-diff-report-tool execution started.
XML parsing is started.
Creation of configuration report is started.
Exception in thread "main" javax.xml.stream.XMLStreamException: Exception was thrown while processing: diff_config.xml
	at com.github.checkstyle.parser.CheckstyleConfigurationsParser.parseConfiguration(CheckstyleConfigurationsParser.java:129)
	at com.github.checkstyle.parser.CheckstyleConfigurationsParser.parse(CheckstyleConfigurationsParser.java:102)
	at com.github.checkstyle.Main.main(Main.java:185)
Caused by: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,2]
```

from https://github.com/checkstyle/checkstyle/actions/runs/4014660497/jobs/6895449715#step:9:4203

--------------------------------------
Successful report generation as proof that workflow functions properly again:

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2cb84e0_2023134216/reports/diff/index.html

cc @romani @rnveach @strkkk @pbludov

File: .ci/validation.sh [hunk 1]
```diff
@@ -1128,6 +1128,22 @@ ci-temp-check)
...
     ;;
 
+check-wildcards-on-pitest-target-classes)
+  ALL_CLASSES=$(xmlstarlet sel \
+    -N x=http://maven.apache.org/POM/4.0.0 \
+    -t -v "/x:project/x:profiles/x:profile//x:targetClasses/x:param" \
+    -n pom.xml)
+
+  CLASSES_NO_WILDCARD=$(echo "$ALL_CLASSES" | grep -v ".*\*\$" | grep -v -e '^\s*$' || echo)
+  CLASSES_NO_WILDCARD_COUNT=$(echo "$CLASSES_NO_WILDCARD" | grep -v -e '^\s*$' | wc -l)
```
Comment (romani) [inline .ci/validation.sh:1138]: Strange that extra grep is required to make count.
CLASSES_NO_WILDCARD it might means that name is not correct.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .ci/validation.sh (hunk 1 above)
Comment (nrmancuso) [reply .ci/validation.sh:1138]: This avoids blank line that is result of previous grep.

File: .ci/validation.sh [hunk 2]
```diff
@@ -1128,6 +1128,22 @@ ci-temp-check)
     exit ${#FILES_NO_CONCURRENCY[@]}
     ;;
 
+check-wildcards-on-pitest-target-classes)
+  ALL_CLASSES=$(xmlstarlet sel \
+    -N x=http://maven.apache.org/POM/4.0.0 \
+    -t -v "/x:project/x:profiles/x:profile//x:targetClasses/x:param" \
```
Comment (rnveach) [inline .ci/validation.sh:1134]: You don't need to specify profile id (starts with `pitest-`) or such to ensure we avoid conflicts?

**Review by rnveach — APPROVED** [review]

File: .ci/validation.sh (hunk 2 above)
Comment (nrmancuso) [reply .ci/validation.sh:1134]: I would rather wait for false positive to appear and update this script than to make this expression more complicated than it has to be. What do you think?

Comment (rnveach) [reply .ci/validation.sh:1134]: I think it makes sense to put in this one ID check.

**Review by romani — APPROVED** [review]

**Pull request merged**